### ISSUE TITLE
Consider increasing max_semi_space_size

### ISSUE THREAD

**Issue by dapplion** [ChainSafe/lodestar#2115]
According to GC metrics on master the beacon node spends 10% of the time doing Scavenge GC runs. 

![Screenshot from 2021-02-28 20-45-51](https://user-images.githubusercontent.com/35266934/109431313-37d48880-7a06-11eb-8560-95d6b0db8f40.png)

If the metrics are correct this article suggests increasing `max_semi_space_size` with the flag `--max_semi_space_size` to reduce the frequency of these runs.

https://www.alibabacloud.com/blog/better-node-application-performance-through-gc-optimization_595119

**Comment by wemeetagain** [comment]
cc @matthewkeil re #5829

**Comment by matthewkeil** [comment]
@wemeetagain I had this set on `beta` when we discovered the memory leak in #5851 so it was hard to see full effect but it did cut the Scavenge time in line with the worker new space adjustment before things started to get wonky when the heap grew from the leak.  It will most definitely be a nice tune-up though and I will get this set correctly once we get the leak addressed and can see the fruits of the change!!

**Comment by matthewkeil** [comment]
As a note, the docs are not super clear about what "semi" space is.  But after digging in the codebase its related to young generation and `kNewLargeObjectSpaceToSemiSpaceRatio` which always equals `1`;

```c++
size_t Heap::YoungGenerationSizeFromSemiSpaceSize(size_t semi_space_size) {
  return semi_space_size * (2 + kNewLargeObjectSpaceToSemiSpaceRatio);
}
```

**Comment by matthewkeil** [comment]
Memory leak #5851 resolved.  Deploying `unstable` to `feat2` with `--max-semi-space-size=64`.  The max scavenge witnessed currently was on mainnet in the unstable group.  All other instances were lower and some substantially lower. 
Unsure how the larger value will affect performance on the smaller instances so will let run for a few days without the worker to see what turns up.

`unstable-mainnet-hzax41`
![Screenshot 2023-08-10 at 1 23 13 PM](https://github.com/ChainSafe/lodestar/assets/18608739/93f1e801-5e15-43a2-b539-154545e26d9a)

`unstable-novc-ctvpss`
![Screenshot 2023-08-10 at 1 25 03 PM](https://github.com/ChainSafe/lodestar/assets/18608739/7cb0297e-2acc-4e31-bcde-b0c717196aef)

`unstable-sm1v-ctvpss`
![Screenshot 2023-08-10 at 1 25 16 PM](https://github.com/ChainSafe/lodestar/assets/18608739/891ad195-341a-4766-8dca-588eb5d0b6ca)

**Comment by matthewkeil** [comment]
## Results 

`feat2` with `--max-semi-space-size=64`
`feat1` with `--max-semi-space-size=128`
`beta` with `--max-semi-space-size=256`

all runs are with `use_worker=false`

### `group-mainnet-hzax41`

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691814600000&to=1691901000000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cfeat2-mainnet-hzax41&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691814600000&to=1691901000000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cfeat1-mainnet-hzax41&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691832600000&to=1691904600000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cbeta-mainnet-hzax41&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

#### `feat2`
![Screenshot 2023-08-12 at 10 05 09 PM](https://github.com/ChainSafe/lodestar/assets/18608739/ba90cc52-79d5-4154-96c1-ecd4e000c76b)

#### `feat1`
![Screenshot 2023-08-12 at 10 53 57 PM](https://github.com/ChainSafe/lodestar/assets/18608739/c23dbc7e-0924-42de-ad16-d0c54a220fae)

#### `beta`
![Screenshot 2023-08-12 at 11 10 09 PM](https://github.com/ChainSafe/lodestar/assets/18608739/925099ff-17b9-46f6-8ad1-e5ade4446004)


### `group-lg1k-hzax41`

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691814600000&to=1691901000000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cfeat2-lg1k-hzax41&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691814600000&to=1691901000000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cfeat1-lg1k-hzax41&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691832600000&to=1691904600000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cbeta-lg1k-hzax41&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

#### `feat2`
![Screenshot 2023-08-12 at 11 22 24 PM](https://github.com/ChainSafe/lodestar/assets/18608739/d8be1309-9f8b-447d-bb70-ed5618dff7f4)

#### `feat1`
![Screenshot 2023-08-12 at 11 22 34 PM](https://github.com/ChainSafe/lodestar/assets/18608739/7115012e-fec4-4303-a5ba-8f7fce59a3de)

#### `beta`
![Screenshot 2023-08-12 at 11 22 45 PM](https://github.com/ChainSafe/lodestar/assets/18608739/207027bb-e07c-4170-98bf-e32fae09071b)


### `feat2-md16-ctvpsm`

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691814600000&to=1691901000000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cfeat2-md16-ctvpsm&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691814600000&to=1691901000000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cfeat1-md16-ctvpsm&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691832600000&to=1691904600000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cbeta-md16-ctvpsm&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

#### `feat2`
![Screenshot 2023-08-12 at 11 23 46 PM](https://github.com/ChainSafe/lodestar/assets/18608739/73303b10-b666-4830-802b-7aa327680260)

#### `feat1`
![Screenshot 2023-08-12 at 11 23 53 PM](https://github.com/ChainSafe/lodestar/assets/18608739/790edad8-aec9-46ab-a1bd-8c802e42d3b4)

#### `beta`
![Screenshot 2023-08-12 at 11 24 01 PM](https://github.com/ChainSafe/lodestar/assets/18608739/b2fe1840-6fa7-46f1-b2f9-360ee1eaa7bb)


### `feat2-sm1v-ctvpss`

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691814600000&to=1691901000000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cfeat2-sm1v-ctvpss&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691814600000&to=1691901000000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cfeat1-sm1v-ctvpss&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

https://grafana-lodestar.chainsafe.io/d/lodestar_vm_host/lodestar-vm-host?from=1691832600000&to=1691904600000&var-DS_PROMETHEUS=default&var-rate_interval=1h&var-Filters=instance%7C%3D%7Cbeta-sm1v-ctvpss&orgId=1&var-beacon_job=$%7BVAR_BEACON_JOB%7D

#### `feat2`
![Screenshot 2023-08-12 at 11 24 34 PM](https://github.com/ChainSafe/lodestar/assets/18608739/d513d4c7-57fb-4be0-ba61-81844099b572)

#### `feat1`
![Screenshot 2023-08-12 at 11 24 40 PM](https://github.com/ChainSafe/lodestar/assets/18608739/449b1088-b991-414a-8a35-02dec6c9feda)

#### `beta`
![Screenshot 2023-08-12 at 11 24 46 PM](https://github.com/ChainSafe/lodestar/assets/18608739/25eec984-7459-488d-804a-8697e358af01)

**Comment by matthewkeil** [comment]
deployed `512mb` to `beta` at timestamp below.  Will pull metrics in a day after it stabilizes to see how they compare to `256mb` that was also on `beta`

![Screenshot 2023-08-16 at 12 00 09 AM](https://github.com/ChainSafe/lodestar/assets/18608739/6e68c772-821e-42a8-bc27-cdb800cdf795)

**Comment by matthewkeil** [comment]
Moving the new space to `512mb` was detrimental.  On mainnet the scavenge dropped but for some reason the mark-and-sweep started to climb considerably.  I am not sure of the phenomena here but it's not worth investigating.  

![Screenshot 2023-08-22 at 5 27 38 AM](https://github.com/ChainSafe/lodestar/assets/18608739/2da5f5aa-2801-400a-b565-8347fa48021f)

The sweet spot for setting new space is at a value similar to the net rate at which objects are created/collected such that they exist in only the from space and do not get moved to the to space (as a net average) when collection occurs.  GC tends to drop a bit further (as a % of CPU time) up to roughly as threshold of two times the rate of object creation.  At this point though the space is so large that it tends to affect performance of the node, likely do to searching for objects during runtime.

Current creation rate on `unstable-mainnet-hzax41`on a 30 day timeline with 7d `$rate_interval` is roughly `150mb`.  To keep the number to an even page interval a setting of `148mb` or `152mb` is recommended.

![Screenshot 2023-08-22 at 5 38 07 AM](https://github.com/ChainSafe/lodestar/assets/18608739/e041df0d-a603-4e12-a6a9-dfa44ed65e0d)

Using a 6h `$rate_interval` to confirm makes those values seem appropriate for a first go.  We can reassess in another couple of weeks after setting to see how things proceed.

![Screenshot 2023-08-22 at 5 42 33 AM](https://github.com/ChainSafe/lodestar/assets/18608739/eb1ff77b-461d-4a78-be88-0bab405090cf)

This value is set at the command line however it is possible to programmatically adjust it during startup from historical data.  The same is true of the new space adjustment for the network worker and assumptions made here apply to the worker as most of the scavenged garbage is network related.  This investigation started with #5829 and that fact was proven out there. As a note once the heap size is set at startup it is not possible to change the value in either case.

**Comment by matthewkeil** [comment]
Final not for reference:

- A value that is too low results in excess scavenge GC time.
- A value that is marginally too high results in node performance degradation from increased variable lookup time
- A value that is very high results in mark-and-sweep collection (unknown reason because not researched further)

To set this value in the future run the node without setting a maxYoungGeneration size and see what the net scavenge collection rate is and set this to the same value.  See example in images above.  That value seems to be near the sweet spot.  If this methodology is refined in the future another note will be added below.

**Comment by matthewkeil** [comment]
Closed via #5829
